### PR TITLE
Switch tooling to npm and WP-CLI, automate creation of JavaScript translation files

### PR DESCRIPTION
**Pull request by nickcernis**
## Summary
This pull request updates tooling to use WP-CLI and npm scripts.

It also automates creation of the new JavaScript translation files that Genesis 3.1 requires for WordPress 5.0+ to fully translate JavaScript strings.

- Removes Grunt in favour of npm and WP-CLI for automation (allows us to automate more and use WP-CLI language tooling without having to maintain and write our own Grunt plugins that do the same).
- Creates the plugin language file with WP-CLI instead of using Grunt (`wp i18n make-pot`).
- Clones Genesis from GitHub. (This is required temporarily to generate JavaScript language files with correct filenames — the `wp i18n make-json` command uses an md5 hash of the JavaScript file contents to generate the script language filename, so it needs to be able to read the Genesis source files.)
- Checks out the Genesis tag given in `genesisTranslations.genesisBranchOrTag` in `package.json` (currently 3.1.0).
- Uses a custom script to download Genesis translations from the StudioPress language server into `genesis-temp/lib/languages`. (I ported this script from `grunt-glotpress`, so the settings should seem familiar under the `glotpress` key in `package.json`.)
- Creates the JSON language files from the downloaded Genesis translations (`wp i18n make-json`).
- Copies all language files from the Genesis `lib/languages` directory to the plugin's `translations` directory.
- Renames JSON language files so that WordPress can find them. (WordPress expects the JSON language files to be prefixed by `genesis-`, but WP-CLI doesn't include this prefix.)
- Deletes the `genesis-temp` folder where Genesis was cloned to tidy things up.

## Requirements

1. Composer: https://getcomposer.org/doc/00-intro.md
2. Node.js: https://nodejs.org/en/download/current/
3. Access to the Genesis GitHub repo as your current git user.

## To test
1. Type `composer install` to install the WP-CLI i18n dependency.
2. Type `npm ci` to install JavaScript dependencies.
3. Set the Genesis tag you would like to use in the `genesisBranchOrTag` value in the package.json file. (Currently 3.1.0. Needs to be a valid branch of the Genesis repo, or one of these tags: https://github.com/studiopress/genesis/tags.)
4. Set the translation server version you would like to use in the glotpress.slug value in the package.json file. (Currently "genesis/genesis-3.1".)
5. Type `npm start`.

When the process completes you should see:

- [ ] Language files in the plugin's `translations` folder, including a large number of JSON files.
- [ ] Translations when activating the plugin and selecting a language in WordPress. This includes translations for the Genesis editor sidebar in Genesis 3.1 for supported languages (e.g. Spanish).

## Notes
- I made it so the download process completely removes files from the `translations` folder before adding new ones. This is to ensure that old files from previous Genesis versions do not remain. If this behaviour is not desired I can adjust it.
- @remkus You may need to adjust your `deploy.sh` script to exclude the new 'bin' and 'vendor' directories ('vendor' is created by composer).
- I have not included the Genesis 3.1 translations in this PR. A separate PR for this is here: https://github.com/remkus/genesis-translations/pull/11.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,25 +1,48 @@
...
-    "grunt-contrib-qunit": "~1.2.0",
-    "grunt-glotpress": "~0.2.2",
-    "grunt-jsvalidate": "~0.2.2",
-    "grunt-wp-i18n": "~0.5.4"
-  }
+	"name": "genesis-translations",
+	"version": "2.10.1",
+	"description": "Genesis Translations",
+	"bugs": {
+		"url": "https://github.com/thememix/genesis-translations/issues"
```
Comment (GaryJones) [inline package.json]: URL is incorrect (owner is `remkus`).

File: package.json [hunk 2]
```diff
@@ -1,25 +1,48 @@
...
+	},
+	"homepage": "https://thememix.com/plugins/genesis-translations/",
+	"repository": {
+		"type": "git",
+		"url": "git+https://github.com/thememix/genesis-translations"
```
Comment (GaryJones) [inline package.json]: URL is incorrect (owner is `remkus`).

File: package.json [hunk 3]
```diff
@@ -1,25 +1,48 @@
...
+	},
+	"author": "ThemeMix",
+	"license": "GPL-3.0+",
```
Comment (GaryJones) [inline package.json]: SPDX identifier is invalid. Should be `GPL-3.0-or-later` (or `GPL-2.0-or-later` for it to be better compatible with Genesis and WP core).

File: composer.json [hunk 4]
```diff
@@ -0,0 +1,10 @@
+{
+	"name": "thememix/genesis-translations",
+	"type": "wordpress-plugin",
```
Comment (GaryJones) [inline composer.json]: Missing the `composer/installers` dev package for this value to be used.

File: bin/download-translations.js [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+// 	}
+// },
+
+const axios = require( 'axios' );
+const fs = require( 'fs' );
+
+const package = require( process.cwd() + '/package.json' );
+const options = package.glotpress;
+
+axios.defaults.headers.common['User-Agent'] = `${package.glotpress.textdomain}`;
```
Comment (GaryJones) [inline bin/download-translations.js]: A UA string that indicated it was from `Genesis Translations` plugin would be more indicative that just `genesis`, right?

File: bin/download-translations.js [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+			options.formats.forEach( format => {
+				translations.downloadSet( set, format );
+			});
+		});
+	},
+
+	/**
+	 * Download the translation file for the given translation set and format.
+	 *
+	 * Example download: https://translate.studiopress.com/global/api/projects/genesis/genesis/en-gb/default/export-translations?format=po
```
Comment (GaryJones) [inline bin/download-translations.js]: Example URL should include `.../genesis/genesis-3.1/...`, not `.../genesis/genesis/...`.

File: bin/download-translations.js [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+			} else {
+				console.log(
+					`Could not fetch file. Response: ${response.status} ${response.statusText}`
+				);
+			}
+		});
+	}
+};
+
+translations.init();
```
Comment (GaryJones) [inline bin/download-translations.js]: May be redundant for a script like this, but should / could this `translations` be exported, and the `translations.init()` be called in another file? Thinking of code being defined alongside side-effects.

File: package.json (hunk 3 above)
Comment (nickcernis) [reply package.json]: Thanks, @GaryJones. I changed it here and in `composer.json` to `GPL-2.0-or-later`. If @remkus wants to use `GPL-3.0-or-later` for somer reason I'm happy to make that change.

Comment (remkus) [reply package.json]: Yeah, that's fine!

File: composer.json (hunk 4 above)
Comment (nickcernis) [reply composer.json]: Thanks, added to `require` based on instructions at https://github.com/composer/installers#example-composerjson-file.

File: bin/download-translations.js (hunk 5 above)
Comment (nickcernis) [reply bin/download-translations.js]: Good call — I changed it to `package.name`, which gives `genesis-translations`.

File: bin/download-translations.js (hunk 7 above)
Comment (nickcernis) [reply bin/download-translations.js]: If this was a package, yes, I agree. We would:

1. Expose the init function with a global name like `downloadTranslations` (or `glotpress` if we planned to extend functionality beyond only downloads).
2. Host the package on npm or GitHub (GitHub now has a [package registry](https://help.github.com/en/articles/about-github-package-registry)).
3. Include the package in dev dependencies.
4. Run the script by name in npm scripts (`downloadTranslations --optional-flags-here`).

We're using `download-tranlsations.js` as an alternative to a shell script that better fits with the other tooling and running it directly only via npm scripts, though. I don't think a refactor that follows the above steps is required here. If we want to make the translation script available more widely later, we could certainly consider it then, though.

**Comment by nickcernis** [comment]
Thanks for the review, @GaryJones! I've made all the changes you suggested, with the exception of https://github.com/remkus/genesis-translations/pull/10#discussion_r317416841. (Rationale for that given in my reply.) I appreciate you taking the time to look at this. ❤️

**Pull request merged**